AU2/plugins/util/ScoreManager.py

```diff
@@ -0,0 +1,58 @@
+import functools
+from collections import defaultdict
+from typing import Dict, List, Set
+
+from AU2.database.model import Event, Assassin
+# TODO: from ??? import calculate_formula
+
+
+class ScoreManager:
+    def __init__(self, assassin_ids: List[str], formula: str = "", bonuses: Dict[str, int] = {}):
+        self.kill_tree: Dict[str, List[str]] = {}
+        self.attempt_counter: Dict[str, int] = {}
+        self.live_assassins = set(assassin_ids)
+        self.formula = formula
+        self.bonuses = bonuses
+
+    def add_event(self, e: Event):
+        # adding an event invalidates the cache
+        self._score.cache_clear()
+        for (killer, victim) in e.kills:
+            self.kill_tree.setdefault(killer, []).append(victim)
+            self.live_assassins.discard(victim)
+        for assassin_id in e.pluginState.get("CompetencyPlugin", {}).get("attempts", []):
+            self.attempt_counter[assassin_id] = self.attempt_counter.get(assassin_id, 0) + 1
+
+    def _protoconkers(self, identifier: str, visited: Set[str]) -> int:
+        if identifier in visited:
+            return 0
+        else:
+            visited.add(identifier)
+        return sum((1+self._protoconkers(v, visited) for v in self.kill_tree.get(identifier, [])))
+
+    def _conkers(self, identifier: str) -> int:
+        return sum((self._protoconkers(v, set()) for v in self.kill_tree.get(identifier, [])))

```

Comment:
This is based on the definition that Jamie gave me which was
> Let TotalKills(N) be a function that is equal to (TotalKills(P) + 1) summed for each person P that person N killed 
>
> Conkers = sum_{kills p} TotalKills(p)

(where I decided to call TotalKills protoconkers)
but maybe this definition is wrong?

Reply:
Yep, I'm pretty sure it used to be the case that Jamie's TotalKills is conkers. It doesn't really matter either way, since the other can just be generated by c+k (or c-k).

Maybe check with Jamie?